src/main/java/twilightforest/client/renderer/entity/RenderTFBlockGoblin.java

```diff
@@ -1,34 +1,129 @@
+	}
+
+	private void renderChain(T goblin, Entity chain, float yaw, float partialTicks, MatrixStack stack, IRenderTypeBuffer buffer, int light) {
+		if (chain != null) {
+			double x = MathHelper.lerp((double) partialTicks, goblin.lastTickPosX, goblin.getX());
+			double y = MathHelper.lerp((double) partialTicks, goblin.lastTickPosY, goblin.getY());
+			double z = MathHelper.lerp((double) partialTicks, goblin.lastTickPosZ, goblin.getZ());
+
+			double chainInX = (chain.getX() - goblin.getX());
+			double chainInY = (chain.getY() - goblin.getY());
+			double chainInZ = (chain.getZ() - goblin.getZ());
+
+			stack.push();
+			IVertexBuilder ivertexbuilder = buffer.getBuffer(this.chainModel.getLayer(textureLoc));
+
+			stack.translate(chainInX, chainInY, chainInZ);
+			float pitch = chain.prevRotationPitch + (chain.rotationPitch - chain.prevRotationPitch) * partialTicks;
+			stack.multiply(Vector3f.POSITIVE_Y.getDegreesQuaternion(180 - MathHelper.wrapDegrees(yaw)));
+			stack.multiply(Vector3f.POSITIVE_X.getDegreesQuaternion(pitch));
+
+			stack.scale(-1.0F, -1.0F, 1.0F);
+			this.chainModel.render(stack, ivertexbuilder, light, OverlayTexture.DEFAULT_UV, 1.0F, 1.0F, 1.0F, 1.0F);
+			stack.pop();
+
+			if (this.renderManager.isDebugBoundingBox() && !chain.isInvisible() && !Minecraft.getInstance().isReducedDebug()) {
+				stack.push();
+				stack.translate(chainInX, chainInY, chainInZ);
+				this.renderMultiBoundingBox(stack, buffer.getBuffer(RenderType.getLines()), chain, 0.25F, 1.0F, 0.0F);
+				stack.pop();
+			}
+		}
+	}
+
+	private void renderMultiBoundingBox(MatrixStack stack, IVertexBuilder builder, Entity entity, float p_229094_4_, float p_229094_5_, float p_229094_6_) {
+		AxisAlignedBB axisalignedbb = entity.getBoundingBox().offset(-entity.getX(), -entity.getY(), -entity.getZ());
+		WorldRenderer.drawBox(stack, builder, axisalignedbb, p_229094_4_, p_229094_5_, p_229094_6_, 1.0F);
+	}
+
+	@Override
+	public boolean shouldRender(T entity, ClippingHelperImpl clippingHelper, double p_225626_3_, double p_225626_5_, double p_225626_7_) {
+		if (super.shouldRender(entity, clippingHelper, p_225626_3_, p_225626_5_, p_225626_7_)) {
+			return true;
+		} else {
+
+			Vec3d vec3d = this.getPosition(entity.block, (double) entity.block.getHeight() * 0.5D, 1.0F);
+			Vec3d vec3d1 = this.getPosition(entity.block, (double) entity.block.getEyeHeight(), 1.0F);
+			return clippingHelper.isVisible(new AxisAlignedBB(vec3d1.x, vec3d1.y, vec3d1.z, vec3d.x, vec3d.y, vec3d.z));
+		}
+	}
+
+	private Vec3d getPosition(Entity p_177110_1_, double p_177110_2_, float p_177110_4_) {
+		double d0 = MathHelper.lerp((double) p_177110_4_, p_177110_1_.lastTickPosX, p_177110_1_.getX());
+		double d1 = MathHelper.lerp((double) p_177110_4_, p_177110_1_.lastTickPosY, p_177110_1_.getY()) + p_177110_2_;
+		double d2 = MathHelper.lerp((double) p_177110_4_, p_177110_1_.lastTickPosZ, p_177110_1_.getZ());
+		return new Vec3d(d0, d1, d2);
```

Comment:
are these methods copied from vanilla? please give the params proper names and add `// [VanillaCopy]` comments as appropriate.